### ISSUE TITLE
Using RDataTracker to resolve relationships among variables

### ISSUE THREAD

**Issue by MKLau** [End-to-end-provenance/eep-usecases#3]
0. I've been developing a model for conducting ecosystem simulations using the pitcher plant system 
1. The model was developed mathematically and then implemented in Mathematica and then translated into R
2. I found myself desiring a flow diagram of the relationships among variables both to check the accuracy of the computational implementation (i.e. debugging) and a representation for explaining the model to others 
3. RDataTracker collects all the necessary information now excepting that it doesn't collect the internals of functions loaded from libraries 
4. DDG explorer and CamFlow provide graphics that are helpful but are not the exact representation that I would like, which is a flow diagram in which variables are related by functions or operations 
5. I've added this to the use-cases, including an abstracted version with three variables. These are in the *src/pplant* directory.

**Comment by MKLau** [comment]
https://raw.githubusercontent.com/End-to-end-provenance/eep-usecases/master/src/pplant/ddge_pplant.png

**Comment by MKLau** [comment]
It's hard to represent what I would like to see exactly for the pitcher plant model, but the generalized version would look like: 

x -> y -> z

Each arrow would be a relation based on the functions or operations that are using and then generating the variables.

**Comment by MKLau** [comment]
Aaron pointed out that this feature is similar to what's provided for different purposes by two existing tools. I haven't used either but we've talked about Kepler in various conversations. Both look worth checking out:

kepler workflow diagrammer
https://kepler-project.org

Stella architect 
https://www.iseesystems.com/store/products/stella-architect.aspx

**Comment by MKLau** [comment]
Connected to this. I've also made an issue for DDG explorer that I'm also copying to CamFlow 

https://github.com/End-to-end-provenance/DDG-Explorer/issues/81

**Comment by MKLau** [comment]
Based on Barb's suggestion, I took a look at the "Show how this value was computed" option. And in the generalized example, it works really well! The following shows the "Derived" plot for the value 3-z.

**Comment by MKLau** [comment]
![screen shot 2017-04-10 at 3 35 24 pm](https://cloud.githubusercontent.com/assets/5342857/24879044/82b1bd0e-1e03-11e7-8b53-a190b2c4c927.png)

![screen shot 2017-04-10 at 3 35 31 pm](https://cloud.githubusercontent.com/assets/5342857/24879050/85993f60-1e03-11e7-89f3-1de466e99dee.png)

**Comment by MKLau** [comment]
This is quite a bit more complicated for the pitcher plant model, but the representation still provides useful insights for me to determine how my variable of interest is being computed by the script. Here's the full ddg followed by the "how it was computed" ddg for 32-x, which you can see as the highlighted circle in the full ddg.

**Comment by MKLau** [comment]
![screen shot 2017-04-10 at 3 49 24 pm](https://cloud.githubusercontent.com/assets/5342857/24879586/63749f7c-1e05-11e7-9458-946a71ed42da.png)

**Comment by MKLau** [comment]
![screen shot 2017-04-10 at 3 50 33 pm](https://cloud.githubusercontent.com/assets/5342857/24879609/7447b730-1e05-11e7-9810-c96c47641e02.png)

**Comment by MKLau** [comment]
A couple of take-homes from this: 

0. The "Show the this was computed" is really close to the feature I requested. It works almost exactly for a general, highly simplified, example. 
1. In a "real" example, things get complicated fast, so the visual really needs to be effective and judicious about what information it shows. 
3. Something to notice in the pitcher plant example is that there are some natural groups that form. It might be useful to look into how these sub-graphs might be detected algorithmically. 
4. Identifying the node that I wanted actually took a bit of exploration. Because of the loops, there are multiple places where a "x" variable appears in the ddg. Being able to quickly select and view the "how computed" plots would be essential for this feature to function in this capacity.

**Comment by margoseltzer** [comment]
> On Apr 10, 2017, at 3:57 PM, MKLau <notifications@github.com> wrote:
> 
> A couple of take-homes from this:
> 
> 	• The "Show the this was computed" is really close to the feature I requested. It works almost exactly for a general, highly simplified, example.
> 	• In a "real" example, things get complicated fast, so the visual really needs to be effective and judicious about what information it shows.
> 	• Something to notice in the pitcher plant example is that there are some natural groups that form. It might be useful to look into how these sub-graphs might be detected algorithmically.

This is exactly the kind of clustering Peter did in Orbiter:
https://www.usenix.org/legacy/event/tapp11/tech/final_files/MackoSeltzer.pdf

- M

**Comment by MKLau** [comment]
Yeah, the graph summarization tool looks like it would do, except that I would want to substitute data nodes for process nodes. 

![screen shot 2017-04-10 at 6 23 00 pm](https://cloud.githubusercontent.com/assets/5342857/24885120/c7cd59a4-1e1a-11e7-866a-9fe6f96eed2f.png)
